File: phpoc-web/src/ledger/summary_policy.js
/**
 * SummaryPolicy — year/month boundary summary block insertion policies.
 *
 * Pluggable policy hierarchy. The default YearMonthSummaryPolicy matches
 * the CLI's original behavior. Summary blocks are inserted between the
 * previous block and the upcoming day block when a date boundary is crossed.
 *
 * Usage:
 *   import { YearMonthSummaryPolicy } from './summary_policy.js';
 *   const policy = new YearMonthSummaryPolicy(crypto, masterKey, identitySecret);
 *   const blocks = policy.getSummaryBlocks(prevBlock, '2026-02-01');
 */

import { getBlockHash } from './utils.js';

// ── Helpers ────────────────────────────────────────────────────────────

function parseDate(dateStr) {
  const [y, m, d] = dateStr.split('-').map(Number);
  return { year: y, month: m, day: d };
}

function formatMonth(year, month) {
  return `${year}-${String(month).padStart(2, '0')}`;
}

/**
 * Build a year_summary block with seal and optional identity signature.
 */
function makeYearSummary(crypto, masterKey, identitySecret, year, prevHash, dateStr) {
  const summary = {
    type: 'year_summary',
    year,
    prev_hash: prevHash,
    date: dateStr,
  };
  const json = JSON.stringify(summary, Object.keys(summary).sort());
  summary.year_hash = crypto.seal(json, masterKey);
  if (identitySecret) {
    summary.signature = crypto.sign(summary.year_hash, identitySecret);
  }
  return summary;
}

/**
 * Build a month_summary block with seal and optional identity signature.
 */
function makeMonthSummary(crypto, masterKey, identitySecret, month, prevHash, dateStr) {
  const summary = {
    type: 'month_summary',
    month,
    prev_hash: prevHash,
    date: dateStr,
  };
  const json = JSON.stringify(summary, Object.keys(summary).sort());
  summary.month_hash = crypto.seal(json, masterKey);
  if (identitySecret) {
    summary.signature = crypto.sign(summary.month_hash, identitySecret);
  }
  return summary;
}

// ── Abstract base ──────────────────────────────────────────────────────

class SummaryPolicy {
  /**
   * @param {object} crypto - CryptoService-like object with seal/sign methods.
   * @param {string} masterKey - Hex master key for sealing.
   * @param {string|null} [identitySecret=null] - Optional identity secret for signing.
   */
  constructor(crypto, masterKey, identitySecret = null) {
    this.crypto = crypto;
    this.masterKey = masterKey;
    this.identitySecret = identitySecret;
  }

  /**
   * Determine which summary blocks to insert between prevBlock and currDateStr.
   * @param {object} prevBlock - The last block currently in the ledger.
   * @param {string} currDateStr - ISO date string of the upcoming day block.
   * @returns {object[]} Array of summary block dicts (may be empty).
   */
  getSummaryBlocks(prevBlock, currDateStr) {
    throw new Error('getSummaryBlocks() not implemented');
  }
}

// ── YearMonthSummaryPolicy ─────────────────────────────────────────────

class YearMonthSummaryPolicy extends SummaryPolicy {
  /**
   * Default policy: inserts year_summary on year boundary and
   * month_summary on month boundary.
   *
   * Correctly handles cross-year month boundaries (e.g., Dec→Feb inserts
   * a month summary for Jan, not just the skipped month).
   */
  getSummaryBlocks(prevBlock, currDateStr) {
    const summaries = [];
    const currDate = parseDate(currDateStr);

    let prevHash = getBlockHash(prevBlock);

    // Resolve the effective previous year and month.
    // For a month_summary block, the 'month' field carries the actual
    // time period (e.g. "2025-12"), which may differ from the 'date' field.
    let prevYear, prevMon;
    if (prevBlock.type === 'month_summary' && prevBlock.month) {
      const parts = prevBlock.month.split('-');
      prevYear = parseInt(parts[0], 10);
      prevMon = parseInt(parts[1], 10);
    } else {
      const prevDate = parseDate(prevBlock.date || '1970-01-01');
      prevYear = prevDate.year;
      prevMon = prevDate.month;
    }

    const yearsDiff = currDate.year - prevYear;

    // Year boundary: insert year summary if year changed and prev
    // block is not already a year_summary
    if (currDate.year > prevYear && prevBlock.type !== 'year_summary') {
      const yearSummary = makeYearSummary(
        this.crypto, this.masterKey, this.identitySecret,
        prevYear, prevHash, currDateStr
      );
      summaries.push(yearSummary);
      prevHash = yearSummary.year_hash;
    }

    // Evaluate whether the effective month has actually changed.
    const monthChanged = (currDate.month !== prevMon || yearsDiff > 0);

    if (monthChanged) {
      // Compute the candidate month to summarize.
      let summarizeYear, summarizeMon;
      if (currDate.month === 1) {
        summarizeYear = currDate.year - 1;
        summarizeMon = 12;
      } else {
        summarizeYear = currDate.year;
        summarizeMon = currDate.month - 1;
      }

      const monthStr = formatMonth(summarizeYear, summarizeMon);

      // Don't insert a month summary if:
      //   a) The previous block is already a month_summary for this month, OR
      //   b) December summary when a year summary was just inserted
      const isSameMonth = (
        prevBlock.type === 'month_summary' &&
        prevBlock.month === monthStr
      );
      if (!isSameMonth) {
        const monthSummary = makeMonthSummary(
          this.crypto, this.masterKey, this.identitySecret,
          monthStr, prevHash, currDateStr
        );
        summaries.push(monthSummary);
      }
    }

    return summaries;
  }
}

// ── YearOnlySummaryPolicy ──────────────────────────────────────────────

class YearOnlySummaryPolicy extends SummaryPolicy {
  /** Only inserts year summaries, never month summaries. */
  getSummaryBlocks(prevBlock, currDateStr) {
    const summaries = [];
    const prevDate = parseDate(prevBlock.date || '1970-01-01');
    const currDate = parseDate(currDateStr);

    if (currDate.year > prevDate.year && prevBlock.type !== 'year_summary') {
      const prevHash = getBlockHash(prevBlock);
      const yearSummary = makeYearSummary(
        this.crypto, this.masterKey, this.identitySecret,
        prevDate.year, prevHash, currDateStr
      );
      summaries.push(yearSummary);
    }

    return summaries;
  }
}

// ── NoSummaryPolicy ────────────────────────────────────────────────────

class NoSummaryPolicy extends SummaryPolicy {
  /** Never inserts any summary blocks. */
  getSummaryBlocks(prevBlock, currDateStr) {
    return [];
  }
}

export {
  SummaryPolicy,
  YearMonthSummaryPolicy,
  YearOnlySummaryPolicy,
  NoSummaryPolicy,
};

File: domain/ledger/summary_policy.py
"""SummaryPolicy: year/month boundary summary block insertion.

Extracts the summary insertion logic from core/ledger.py sync_day() into
a pluggable policy hierarchy. The default YearMonthSummaryPolicy matches
the original behavior exactly.
"""

from domain.ledger.helpers import get_block_hash

import json
import time
from abc import ABC, abstractmethod
from typing import Optional, List, Dict, Any

from security.crypto import AbstractCryptoManager


class SummaryPolicy(ABC):
    """Abstract base for summary block policies.

    Subclasses implement get_summary_blocks() which returns a list of
    summary block dicts (year_summary, month_summary) to insert between
    the previous block and the upcoming day block.
    """

    def __init__(
        self,
        crypto: AbstractCryptoManager,
        identity_secret: Optional[bytes] = None,
    ):
        self.crypto = crypto
        self.identity_secret = identity_secret

    @abstractmethod
    def get_summary_blocks(
        self, prev_block: dict, curr_date_str: str
    ) -> List[Dict[str, Any]]:
        """Determine which summary blocks to insert between prev_block and curr_date_str.

        Args:
            prev_block: The last block currently in the ledger (any type).
            curr_date_str: ISO date string of the upcoming day block.

        Returns:
            List of summary block dicts to insert (may be empty).
        """
        ...

    def _make_year_summary(
        self, year: int, prev_hash: str, date_str: str
    ) -> Dict[str, Any]:
        """Build a year_summary block with seal and optional identity signature."""
        summary = {
            "type": "year_summary",
            "year": year,
            "prev_hash": prev_hash,
            "date": date_str,
        }
        summary["year_hash"] = self.crypto.seal(
            json.dumps(summary, sort_keys=True)
        )
        if self.identity_secret:
            summary["signature"] = self.crypto.sign(
                summary["year_hash"], self.identity_secret
            )
        return summary

    def _make_month_summary(
        self, month: str, prev_hash: str, date_str: str
    ) -> Dict[str, Any]:
        """Build a month_summary block with seal and optional identity signature."""
        summary = {
            "type": "month_summary",
            "month": month,
            "prev_hash": prev_hash,
            "date": date_str,
        }
        summary["month_hash"] = self.crypto.seal(
            json.dumps(summary, sort_keys=True)
        )
        if self.identity_secret:
            summary["signature"] = self.crypto.sign(
                summary["month_hash"], self.identity_secret
            )
        return summary

    @staticmethod
    def _parse_date(date_str: str) -> time.struct_time:
        """Parse an ISO date string into a time.struct_time."""
        return time.strptime(date_str, "%Y-%m-%d")

    @staticmethod
    def _format_month(year: int, month: int) -> str:
        """Format a (year, month) pair as 'YYYY-MM'."""
        return f"{year}-{month:02d}"


class YearMonthSummaryPolicy(SummaryPolicy):
    """Default policy: inserts year_summary on year boundary and
    month_summary on month boundary.

    Correctly handles cross-year month boundaries (e.g., Dec→Feb inserts
    a month summary for Jan, not just Dec's month).
    """

    def get_summary_blocks(
        self, prev_block: dict, curr_date_str: str
    ) -> List[Dict[str, Any]]:
        summaries: List[Dict[str, Any]] = []
        curr_date = self._parse_date(curr_date_str)

        prev_hash = get_block_hash(prev_block)

        # Resolve the effective previous year and month.
        # For a month_summary block, the 'month' field carries the actual
        # time period (e.g. "2025-12"), which may differ from the 'date'
        # field (which is set to the upcoming day's date for prev_hash
        # continuity).
        if prev_block.get("type") == "month_summary" and "month" in prev_block:
            month_parts = prev_block["month"].split("-")
            prev_year = int(month_parts[0])
            prev_mon = int(month_parts[1])
        else:
            prev_date = self._parse_date(prev_block.get("date", "1970-01-01"))
            prev_year = prev_date.tm_year
            prev_mon = prev_date.tm_mon

        years_diff = curr_date.tm_year - prev_year

        # Year boundary: insert year summary if year changed and prev
        # block is not already a year_summary
        if (
            curr_date.tm_year > prev_year
            and prev_block.get("type") != "year_summary"
        ):
            year_summary = self._make_year_summary(
                prev_year, prev_hash, curr_date_str
            )
            summaries.append(year_summary)
            # Update prev_hash for subsequent summaries
            prev_hash = year_summary["year_hash"]

        # Evaluate whether the effective month has actually changed.
        # Month changed if:
        #   - Different month within the same year, OR
        #   - Years are different (month necessarily changed)
        month_changed = (
            curr_date.tm_mon != prev_mon or years_diff > 0
        )

        if month_changed:
            # Compute the candidate month to summarize.
            # This is the month just before the current date, accounting for
            # year wrap (Jan → previous December).
            if curr_date.tm_mon == 1:
                summarize_year = curr_date.tm_year - 1
                summarize_mon = 12
            else:
                summarize_year = curr_date.tm_year
                summarize_mon = curr_date.tm_mon - 1

            month_str = self._format_month(summarize_year, summarize_mon)

            # Don't insert a month summary if:
            #   a) The previous block is already a month_summary for this month, OR
            #   b) This would be December summary when a year summary was just
            #      inserted (year summary already covers December)
            is_same_month = (
                prev_block.get("type") == "month_summary"
                and prev_block.get("month") == month_str
            )
            dec_with_year = (
                summarize_mon == 12
                and years_diff > 0
                and any(s.get("type") == "year_summary" for s in summaries)
            )

            if not is_same_month and not dec_with_year:
                month_summary = self._make_month_summary(
                    month_str,
                    prev_hash,
                    curr_date_str,
                )
                summaries.append(month_summary)

        return summaries


class YearOnlySummaryPolicy(SummaryPolicy):
    """Policy that only inserts year summaries, never month summaries."""

    def get_summary_blocks(
        self, prev_block: dict, curr_date_str: str
    ) -> List[Dict[str, Any]]:
        summaries: List[Dict[str, Any]] = []
        prev_date = self._parse_date(prev_block.get("date", "1970-01-01"))
        curr_date = self._parse_date(curr_date_str)

        if (
            curr_date.tm_year > prev_date.tm_year
            and prev_block.get("type") != "year_summary"
        ):
            prev_hash = get_block_hash(prev_block)
            year_summary = self._make_year_summary(
                prev_date.tm_year, prev_hash, curr_date_str
            )
            summaries.append(year_summary)

        return summaries


class NoSummaryPolicy(SummaryPolicy):
    """Policy that never inserts any summary blocks."""

    def get_summary_blocks(
        self, prev_block: dict, curr_date_str: str
    ) -> List[Dict[str, Any]]:
        return []
